File: python_solution/problem22.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Wed Sep 18 10:12:43 2024

@author: sharminalina
"""



class Problem22:
    def GenerateParenthesis(self, n: int)->list:
        result = []
        self.generate(result,'',0,0,n)
        return result
    def generate(self, result: list, current: str, opening_pos: int, closing_pos: int, max_pos: int)-> None:
        if(len(current) == max_pos*2):
            result.append(current)
            return
        if(opening_pos < max_pos):
            self.generate(result, current+'(', opening_pos+1,closing_pos,max_pos)
        if(closing_pos < opening_pos):
            self.generate(result, current+')', opening_pos,closing_pos+1,max_pos)
            
        
        
    
def main():
    # Problem 22: Give a number to generate the valid possible parenthesis
    n = int(input("Enter a number to generate the valid parenthesis: "))
    
    
    # Genrate parenthesis
    
    solution_prob22 = Problem22()
    result = solution_prob22.GenerateParenthesis(n)
    
    # Print the list of int whose sum is 0
    print(f"The list of possible parenthesis are : {result}")

if __name__ == "__main__":
    main()

File: C#Solutions/problem22.cs
using System;
using System.ComponentModel;
using System.ComponentModel.DataAnnotations;
using System.Xml.XPath;

namespace AllProblems
{
    public static class Problem22
    {
        public static void Solve()
        {
            Console.WriteLine("Solution to Problem 22: Generate Parenthesis");
            
            Console.WriteLine("Give a number to generate possible parenthesis:");
            int n = int.Parse(Console.ReadLine());
            List<string> result = GenerateParenthesis(n);
            foreach (var item in result)
            {
                Console.WriteLine(item);
            }
            
            
        }

        

        public static List<string> GenerateParenthesis(int n)
        {
            List<string> result = new List<string>();
            generate(result, "", 0, 0, n);
            return result;

        }

        public static void generate(List<string> result, string current, int open, int close, int max ){
            if(current.Length == max*2){
                result.Add(current);
                return;
            }
            if(open < max){
                generate(result, current+'(', open +1, close, max);
            }
            if(close < open){
                generate(result, current+')', open, close+1,max);
            }
        }

        
    }
}
